File: Week_05/Practicum/Shortest Unsorted Continuous Subarray.cpp
// Виж README.md за описание на решението.
class Solution {
public:
    int findUnsortedSubarray(vector<int>& nums) {
        std::stack<int> min, max; // can be done without stacks, just two ints :D
        max.push(INT_MIN);
        min.push(INT_MAX);
        int n = nums.size(), min_idx = 0, max_idx = 0;

        for (int k = 0; k < n; ++k) {
            if (nums[k] < max.top()) {
                max_idx = k;
            } else {
                max.push(nums[k]);
            }

            if (nums[n - 1 - k] > min.top()) {
                min_idx = n - 1 - k;
            } else {
                min.push(nums[n - 1 - k]);
            }
        }

        return max_idx > min_idx ? max_idx - min_idx + 1 : 0;
    }
};

File: Week_05/Practicum/python/Shortest Unsorted Continuous Subarray.py
class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        low, high = len(nums) - 1, 0

        # monotonically increasing stack
        stack = []    
        for i, num in enumerate(nums):
            while stack and stack[-1][1] > num:
                ind, last = stack.pop()
                low = min(low, ind)

            stack.append((i, num))

        # monotonically decreasing stack
        stack = []       
        for i, num in enumerate(reversed(nums)):
            while stack and stack[-1][1] < num:
                ind, last = stack.pop()
                high = max(high, ind)

            stack.append((len(nums) - 1 - i, num))

        # low >= high means array is already sorted
        return 0 if low >= high else high - low + 1